alibi/explainers/permutation_importance.py

```diff
@@ -0,0 +1,659 @@
+            loss_orig[loss_name] = self._compute_loss(y_true=y,
+                                                      y_pred=self.predictor(X),
+                                                      loss_fn=loss_fn,
+                                                      sample_weight=sample_weight)
+
+        # compute permutation feature importance for every feature
+        # TODO: implement parallel version - future work as it can be done for ALE too
+        individual_feature_importance = []
+
+        for ifeatures in tqdm(features, disable=not self.verbose):
+            individual_feature_importance.append(
+                self._compute_permutation_importance(
+                    X=X,
+                    y=y,
+                    loss_fns=loss_fns,
+                    method=method,
+                    kind=kind,
+                    n_repeats=n_repeats,
+                    sample_weight=sample_weight,
+                    features=ifeatures,
+                    loss_orig=loss_orig,
+                )
+            )
+
+        # update meta data params
+        self.meta['params'].update(feature_names=feature_names,
+                                   method=method,
+                                   kind=kind,
+                                   n_repeats=n_repeats,
+                                   sample_weight=sample_weight)
+
+        # build and return the explanation object
+        return self._build_explanation(feature_names=feature_names,  # type: ignore[arg-type]
+                                       individual_feature_importance=individual_feature_importance)
+
+    def _compute_loss(self,
+                      y_true: np.ndarray,
+                      y_pred: np.ndarray,
+                      loss_fn: Callable[[np.ndarray, np.ndarray, Optional[np.ndarray]], float],
+                      sample_weight: Optional[np.ndarray]) -> float:
+        """
+        Helper function to compute the loss. It also checks if the loss function expects the arguments `y_true`,
+        `y_pred`, and optionally `sample_weight`.
+
+        Parameters
+        ----------
+        y_true
+            Ground truth targets.
+        y_pred
+            Predicted outcome as returned by the classifier.
+        loss_fn
+            Loss function to be used. Note that the loss function must be compatible with the `y_true`, `y_pred`, and
+            optionally with `sample_weight`.
+        sample_weight
+            Weight of each sample instance.
+
+        Returns
+        -------
+        Loss value.
+        """
+        # get scoring function arguments
+        args = inspect.getfullargspec(loss_fn).args
+
+        if 'y_true' not in args:
+            raise ValueError('The `scoring` function must have the argument `y_true` in its definition.')
+
+        if 'y_pred' not in args:
+            raise ValueError('The `scoring` function must have the argument `y_pred` in its definition.')
```

Comment:
The alternative would be to loose IMO. We would have the rely on the ordering of the parameters which might lead to errors. For example there are loss functions that are not symmetric in their arguments (e.g. KL). If the ordering of the arguments does not match our expectation this can lead to incorrect results.

Reply:
ok I can see the advantage here, happy to leave this validation in. Since we'll provide built-in loss functions/metrics most people would be able to use those with no effort and custom ones would still be supported, though people may need to write a short wrapper around any existing function to ensure the argument names are as expected by `alibi`.